### DISCUSSION TITLE
EXR's ImageOutput  appendsubimage

### DISCUSSION THREAD

**Discussion by colaub** [OpenImageIO/oiio#4018] [Q&A] [answered]
Hello, 
I am in trouble with my code because oiio is telling me that EXR doesn't support "appendsubimage" whereas if I followed the doc and examples it should... 
Here is my very simple code : 
```c++
bool
writeMultiLayerEXR(const std::string& filename)
{
    std::unique_ptr<OIIO::ImageOutput> out = OIIO::ImageOutput::create(filename);
    
    if (!out->supports("appendsubimage")) {
        DEBUG("Does not support appending of subimages\n");
        return false;
    }
    return true;
}
```
And it return false... Do you have any idea why? What do I miss? 

If I write a single EXR with `ImageOutput` and check its header, it's cleary a EXR2.0. 

Thanks for the help. 

Colin.

**Comment by DartinMavies** [comment, marked as answer]
I think your issue is that `out->supports("appendsubimage")` returns 0 if the image format needs to have to the number of subimages predeclared. From the docs:

```
"appendsubimage" : Does this format support multiple subimages that can be successively appended at will via open(name,spec,AppendSubimage)? A value of 0 means that the format requires pre-declaring the number and specifications of the subimages when the file is first opened, with open(name,subimages,specs).
```

**Reply by colaub** [reply]
:thinking: interesting... So my check could be set too early, I've got something to look for again.
Thanks for the help! 

But that brings me to the original problem. This current code doesn't work either, even if I delete this check : 
```c++
bool
writeMultiLayerEXR(const std::vector<std::string>& layerNames,
                   const std::vector<Image>& images,
                   const std::string& filename)
{
    if (layerNames.size() != images.size()) return false;

    std::unique_ptr<OIIO::ImageOutput> out = OIIO::ImageOutput::create(filename);
    
    //if (images.size() > 1 && !out->supports("appendsubimage")) {
    //     DEBUG("Does not support appending of subimages\n");
    //     return false;
    //}

    for (int i = 0; i < images.size(); ++i) {
        OIIO::ImageSpec newspec = images[i].imageBuf.spec();
        newspec.attribute("oiio:subimagename", layerNames[i]);
        newspec.channelnames.clear();
        newspec.channelnames.assign({layerNames[i]+".R", layerNames[i]+".G", layerNames[i]+".B", layerNames[i]+".A"});

        if (i == 0) {
            if (!out->open(filename, newspec)) {
                images[i].imageBuf.error(out->geterror());
                return false;
            }
        } else {
            if (!out->open(filename, newspec, OIIO::ImageOutput::AppendSubimage)) {
                images[i].imageBuf.error(out->geterror());
                DEBUG("Write Subimage {} failed error : {}", layerNames[i], images[i].imageBuf.geterror());
                return false;
            }
        }
        if (!images[i].imageBuf.write(out.get())) {
            return false;
        }
    }
    out->close();
    DEBUG("Close the image");
    return true;
}

```

This fail during the `out->open(filename, newspec, OIIO::ImageOutput::AppendSubimage` 

Maybe I am miss leading on how to write a list of imageBuf into a multipart EXR ? ...

**Reply by colaub** [reply]
oky I read again more carefully the example comments : 
"Some image formats allow you to do the initial open() call without declaring the specifics of the subimages, and simply append subimages as you go. You can detect this by checking" 
So I guess EXR doesn't allow us, I have to find out how declare subimages before `open`

**Reply by lgritz** [reply]
Yeah, that's it exactly. There is a version of the open() call that takes a list of ImageSpec's up front, one per subimage/part.

We should clarify terminology here.

"Parts" is what OpenEXR calls a file that contains multiple separate images (this is what OpenImageIO calls "subimages"). This was added in OpenEXR 2.0. 

"Layers" is an old OpenEXR 1.x trick (before parts were supported) where there is ONE image in the file, but LOTS of channels, and the channels are named "foo.R", "foo.G", "foo.B", "bar.R", "bar.G", "bar.B", etc. By convention, it is understood that all the channels that start with the same prefix ("foo" or "bar", say) constitute a "layer" that goes together and conceptually forms a separate logical image, even though OpenEXR 1.x only supports one image per file.

**Reply by colaub** [reply]
Ok nice! thanks for the clarification. 
I found out why I missed the info in the last example, the `open` is after the "appendsubimages" check, so I didn't pay attention to the signature. 
Last example code in https://openimageio.readthedocs.io/en/v2.5.4.0/imageoutput.html#multi-image-files 

I will modify my code and hope it works. 
 
  Thanks all!

**Reply by colaub** [reply]
I have a working EXR multipart, thanks! 
Question, is it still possible to create an old "Layers" EXR with OIIO?

**Reply by lgritz** [reply]
"Layers" is not an actual feature of OpenEXR, it's just a convention for how you name the channels (and a prayer that whoever is reading the files your produce also follow the same convention). So, for example, if you have a single-part image that has 6 channels named "diffuse.R", "diffuse.G", "diffuse.B", "specular.R", "specular.G", "specular.B", then congratulations, you have two layers called "diffuse" and "specular". That's it, that's all there is to it.
